### PR TITLE
Proovide instances for Kleisli

### PR DESCRIPTION
**Pull request by AndreasKostler**
Provide `ConcurrentEffect` and `Effect` instances for `Kleisli`

### CODE REVIEW

File: core/shared/src/main/scala/cats/effect/ConcurrentEffect.scala [hunk 1]
```diff
@@ -96,6 +96,13 @@ object ConcurrentEffect {
   implicit def catsWriterTConcurrentEffect[F[_]: ConcurrentEffect, L: Monoid]: ConcurrentEffect[WriterT[F, L, ?]] =
     new WriterTConcurrentEffect[F, L] { def F = ConcurrentEffect[F]; def L = Monoid[L] }
 
+  /**
+   * [[ConcurrentEffect]] instance built for `cats.data.WriterT` values initialized
```
Comment (kubukoz) [inline core/shared/src/main/scala/cats/effect/ConcurrentEffect.scala]: Maybe better with `WriterT` changed to `Kleisli`

**Comment by djspiewak** [comment]
So imagine these instances in evaluating a stream concurrently. Using fs2 to give you an idea:

```scala
val str1: Stream[Kleisli[IO, R, ?], A] = ???
val str2: Stream[Kleisli[IO, R, ?], A] = ???

val results: Kleisli[IO, R, Vector[A]] = 
  str1.merge(str2).compile.toVector

results.run(valueOfR)
```

In order for this to work, the zero of the `Monoid` for the reader is going to have to be repeatedly injected as the `Kleisli` is evaluated over and over again. In other words, most of the suspended `ask` effects are not going to receive `valueOfR`, but are instead going to receive `Monoid[R].empty`! This is really surprising behavior.

It is for this reason that I never defined `Effect` or `ConcurrentEffect` for `Kleisli` (or `StateT`). Even though it's well-defined, the semantics are extremely confusing for users.

**Comment by johnynek** [comment]
When we are using Monoid just to get some value but not in relation to combine I am skeptical.

I think that is the root of the issue here.

**Comment by AndreasKostler** [comment]
Not a good idea - as discussed.

**Pull request closed without merging**